File: 20241015/heemin/두_큐_합_같게_만들기.cpp
#include <string>
#include<iostream>
#include <vector>

using namespace std;

int solution(vector<int> queue1, vector<int> queue2) {
    int answer = -2;
    
    long long sum1=0;
    long long sum2=0;
    
    for(int i=0;i<queue1.size();i++){
        sum1+=queue1[i];
    }
    for(int i=0;i<queue2.size();i++){
        sum2+=queue2[i];
    }
    
    int sum=sum1+sum2;
    
    if(sum % 2 !=0){
        return -1;
    }
    
    int cnt=0;
    int index1=0;
    int index2=0;
    
    int size=queue1.size()+queue2.size();
    while(cnt <= size*2){
        if(sum1 > sum2){
            int num=queue1[index1];
            sum1-=num;
            sum2+=num;
            queue2.push_back(num);
            index1++;
            cnt++;
        }
        else if(sum1 < sum2 ){
            int num=queue2[index2];
            sum1+=num;
            sum2-=num;
            queue1.push_back(num);
            index2++;
            cnt++;
        }
        else{
            break;
        }
    }
    
    sum1==sum2 ? answer=cnt : answer=-1;

    return answer;
}

File: 20241015/jueun/두큐합같게만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        int len = queue1.length + queue2.length;
        int[] arr = new int[len];
        
        //두 큐의 각각의 합, 총합/2
        long q1=0, q2=0, goal=0;
        
        for(int i=0; i<queue1.length; i++) {
            q1 += queue1[i];
            arr[i] = queue1[i];
        }
        
        for(int i=0; i<queue2.length; i++) {
            q2 += queue2[i];
            arr[queue1.length + i] = queue2[i]; 
        }

        goal = (q1 + q2) / 2;
        
        int start = 0, end = queue1.length-1;
        int answer = 0;
        
        while(true) {      
            if(q1 == goal) {
                return answer; 
            }
            else if(q1 < goal) {
                end ++;
                
                if(end >= len)
                    return -1;
                
                q1 += arr[end];
                
            } else {
                q1 -= arr[start];
                
                start ++;
            }
            
            answer++;
        }
    }
}
